### PR TITLE
fix!: single file database

### PR DESCRIPTION
**Pull request by kirisaki-vk**

### CODE REVIEW

File: src/test/java/com/harena/api/service/PatrimoineServiceIT.java [hunk 1]
```diff
@@ -49,10 +48,8 @@ void patrimoine_service() throws IOException {
 
   @Test
   void patrimoine_service_test_with_no_files() throws IOException {
-    String randomFilename = UUID.randomUUID() + "-harena-tests";
-    when(provider.getBucket()).thenReturn(localBucketComponent);
-    when(localBucketComponent.download(any()))
-        .thenReturn(new File("/tmp/random-file-harena_api_test" + randomFilename));
+    when(bucketComponent.download(any()))
+        .thenReturn(Files.createTempFile("harena-tests", null).toFile());
```
Comment (hei-teacher) [inline src/test/java/com/harena/api/service/PatrimoineServiceIT.java]: use static import

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 2]
```diff
@@ -1,72 +1,106 @@
...
-    String id = StringNormalizer.apply(patrimoineName);
-    File patrimoineFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!patrimoineFile.exists()) {
-      return Optional.empty();
+  public Patrimoine getPatrimone(String patrimoineName) {
+    File patrimoineFile;
+    try {
+      patrimoineFile =
+          bucketComponent.download(PATRIMOINE_KEY_PREFIX + StringNormalizer.apply(patrimoineName));
+    } catch (Exception e) {
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: Preciser le type de l'exception qui correspond au NotFound. En particulier, pas TOUTES les exceptions sont des NotFound (eg: erreurs transcientes comme la perte de connection internet)

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 3]
```diff
@@ -1,72 +1,106 @@
...
 
-  @SneakyThrows
   public List<Patrimoine> savePatrimoines(List<Patrimoine> toSavePatrimoines) {
-    File oldPatrimoinesFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!oldPatrimoinesFile.exists()) {
-      oldPatrimoinesFile.createNewFile();
-      HashMap<String, Patrimoine> emptyPatrimoine = new HashMap<>();
-      Files.writeString(oldPatrimoinesFile.toPath(), serializer.serialise(emptyPatrimoine));
+    try {
+      Path tempDir = Files.createTempDirectory("harena-api-toSavePatrimoines");
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: import static

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 4]
```diff
@@ -1,72 +1,106 @@
...
     }
 
-    HashMap<String, Patrimoine> patrimoine =
-        serializer.deserialise(Files.readString(patrimoineFile.toPath()));
-    return Optional.ofNullable(patrimoine.get(id));
+    try {
+      return serializer.deserialise(Files.readString(patrimoineFile.toPath()));
+    } catch (IOException | ClassCastException e) {
+      throw new InternalServerErrorException(
+          String.format("Could not read values of patrimoine %s", patrimoineName));
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: Use String template

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 5]
```diff
@@ -1,72 +1,106 @@
...
-  @SneakyThrows
   public List<Patrimoine> savePatrimoines(List<Patrimoine> toSavePatrimoines) {
-    File oldPatrimoinesFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!oldPatrimoinesFile.exists()) {
-      oldPatrimoinesFile.createNewFile();
-      HashMap<String, Patrimoine> emptyPatrimoine = new HashMap<>();
-      Files.writeString(oldPatrimoinesFile.toPath(), serializer.serialise(emptyPatrimoine));
+    try {
+      Path tempDir = Files.createTempDirectory("harena-api-toSavePatrimoines");
+      toSavePatrimoines.forEach((patrimoine) -> writeFile(tempDir, patrimoine));
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: superfluous ()

Comment (hei-teacher) [reply src/main/java/com/harena/api/service/PatrimoineService.java]: Deep implications for the courageous ones: (1) partial application of function (2) curryfication

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 6]
```diff
@@ -1,72 +1,106 @@
...
 
-    String oldPatrimoineString = Files.readString(oldPatrimoinesFile.toPath());
-    HashMap<String, Patrimoine> patrimoines = serializer.deserialise(oldPatrimoineString);
-    toSavePatrimoines.forEach(
-        patrimoine -> {
-          patrimoines.put(StringNormalizer.apply(patrimoine.nom().toLowerCase()), patrimoine);
-        });
+  public List<Patrimoine> getAllPatrimoine(PageRequest pageRequest) {
+    List<File> patrimoineFiles = new ArrayList<>();
+    ListObjectsV2Request lsRequest =
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: var

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 7]
```diff
@@ -1,72 +1,106 @@
...
+            .build();
 
-    String newPatrimoineString = serializer.serialise(patrimoines);
-    Path tmpFile = Files.createTempFile(PATRIMOINE_KEY, null);
-    File serializedPatrimoine = Files.writeString(tmpFile, newPatrimoineString).toFile();
-    bucketComponent.upload(serializedPatrimoine, PATRIMOINE_KEY);
-    return toSavePatrimoines;
+    ListObjectsV2Iterable lsIterable = bucketConf.getS3Client().listObjectsV2Paginator(lsRequest);
+
+    int pageNumber = 1;
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java:81]: Use it, and increment it!

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 8]
```diff
@@ -1,72 +1,106 @@
...
+    for (ListObjectsV2Response objects : lsIterable) {
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java:82]: Really not sure this is the proper way to do pagination in S3...

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 9]
```diff
@@ -1,72 +1,106 @@
...
+      if (pageRequest.pageNumber() == pageNumber) {
+        objects
+            .contents()
+            .forEach(
+                file -> {
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: superfluous {}

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 10]
```diff
@@ -1,72 +1,106 @@
...
+    }
+
+    return patrimoineFiles.stream()
+        .map(
+            file -> {
+              try {
+                return serializer.deserialise(Files.readString(file.toPath()));
+              } catch (IOException e) {
+                throw new InternalServerErrorException(
+                    String.format("Cannot read %s", file.getName()));
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: String template

File: src/main/java/com/harena/api/service/PatrimoineService.java [hunk 11]
```diff
@@ -1,72 +1,106 @@
...
   }
 
-  @SneakyThrows
-  public List<Patrimoine> getAllPatrimoine() {
-    File patrimoinesFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!patrimoinesFile.exists()) {
-      return List.of();
+  private void writeFile(Path parentDir, Patrimoine patrimoine) {
+    try {
+      Path filePath = Files.createFile(parentDir.resolve(StringNormalizer.apply(patrimoine.nom())));
```
Comment (hei-teacher) [inline src/main/java/com/harena/api/service/PatrimoineService.java]: static import

File: src/test/java/com/harena/api/service/PatrimoineServiceIT.java [hunk 12]
```diff
@@ -24,35 +26,22 @@
...
-    HashMap<String, Patrimoine> baseMap = new HashMap<>();
-    baseMap.put("test", patrimoineZetySupplier.get());
-    Files.writeString(tempFile, serializer.serialise(baseMap));
+    Files.writeString(tempFile, serializer.serialise(patrimoineZetySupplier.get()));
     when(bucketComponent.download(any())).thenReturn(tempFile.toFile());
     when(bucketComponent.upload(any(), any())).thenReturn(new FileHash(NONE, "hash"));
 
-    assertFalse(subject.getAllPatrimoine().isEmpty());
-    assertTrue(subject.getPatrimone("test").isPresent());
+    assertFalse(subject.getAllPatrimoine(PageRequest.of(1, 1)).isEmpty());
```
Comment (hei-teacher) [inline src/test/java/com/harena/api/service/PatrimoineServiceIT.java]: Page(1,1) is not representative of reality. People usually use something like Page(4,10)

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/test/java/com/harena/api/service/PatrimoineServiceIT.java (hunk 1 above)
Comment (MiiJack) [reply src/test/java/com/harena/api/service/PatrimoineServiceIT.java]: Resolved

**Pull request merged**